src/ws-api.ts

```diff
@@ -0,0 +1,97 @@
+import type { Dwn } from '@tbd54566975/dwn-sdk-js';
+
+import { Server } from 'http';
+
+import { WebSocket } from 'ws';
+import { v4 as uuidv4 } from 'uuid';
+import { WebSocketServer } from 'ws';
+
+import { jsonRpcApi } from './json-rpc-api.js';
+import { createJsonRpcErrorResponse, JsonRpcErrorCodes, JsonRpcResponse } from './lib/json-rpc.js';
+
+const SOCKET_ISALIVE_SYMBOL = Symbol('isAlive');
+
+export class WsApi {
+  wsServer: WebSocketServer;
+  dwn: Dwn;
+
+  constructor(server: Server, dwn: Dwn) {
+    this.dwn = dwn;
+    this.wsServer = new WebSocketServer({ server: server });
+  }
+
+  get address() {
+    return this.wsServer.address;
+  }
+
+  listen() {
+    const dwn = this.dwn;
+    this.wsServer.on('connection', function (socket: WebSocket, _request, _client) {
+      socket[SOCKET_ISALIVE_SYMBOL] = true;
+
+      // Pong messages are automatically sent in response to ping messages as required by
+      // the websocket spec. So, no need to send explicit pongs from browser
+      socket.on('pong', function() {
+        this[SOCKET_ISALIVE_SYMBOL] = true;
+      });
+
+      socket.on('message', async function(dataBuffer) {
+        let dwnRequest;
+
+        try {
+          // deserialize bytes into JSON object
+          dwnRequest = dataBuffer.toString();
+          if (!dwnRequest) {
+            const jsonRpcResponse = createJsonRpcErrorResponse(uuidv4(),
+              JsonRpcErrorCodes.BadRequest, 'request payload required.');
+
+            const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
+            return socket.send(responseBuffer);
+          }
+
+          dwnRequest = JSON.parse(dwnRequest);
+        } catch(e) {
+          const jsonRpcResponse = createJsonRpcErrorResponse(
+            uuidv4(), JsonRpcErrorCodes.BadRequest, e.message);
+
+          const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
+          return socket.send(responseBuffer);
+        }
+
+        const { jsonRpcResponse } = await jsonRpcApi.handle(dwnRequest, { transport: 'ws', dwn });
+
+        const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
+        return socket.send(responseBuffer);
+      });
+    });
+
+    // Sometimes connections between client <-> server can get borked in such a way that
+    // leaves both unaware of the borkage. ping messages can be used as a means to verify
+    // that the remote endpoint is still responsive. Server will ping each socket every 30s
+    // if a pong hasn't received from a socket by the next ping, the server will terminate
+    // the socket connection
+    const heartbeatInterval = setInterval(function () {
```

Comment:
When I run the server, every 30 seconds I see this error printed to the console:
```shell
ncaught exception: TypeError: Cannot read properties of undefined (reading 'clients')
    at Timeout._onTimeout (file:///Users/frankhinek/Code/dwn-server/dist/src/ws-api.js:53:39)
    at listOnTimeout (node:internal/timers:573:17)
    at process.processTimers (node:internal/timers:514:7)
```
Issue appears to be that `this` is the scope of the function and not of the class instance.

Think it should be an arrow function:
`const heartbeatInterval = setInterval(() => {`